src/quorumnet/sn_network.cpp

```diff
@@ -0,0 +1,1019 @@
+                status_text = "Invalid quorum connection authentication mechanism: " + (std::string) mech;
+            }
+            else if (frames.size() != 7) {
+                SN_LOG(error, "Bad ZAP authentication request: invalid request message size");
+                status_code = "500";
+                status_text = "Invalid CURVE authentication request\n";
+            }
+            else if (frames[6].size() != 32) {
+                SN_LOG(error, "Bad ZAP authentication request: invalid request pubkey");
+                status_code = "500";
+                status_text = "Invalid public key size for CURVE authentication";
+            }
+            else {
+                std::string ip{view(frames[3])}, pubkey{view(frames[6])};
+                if (allow_connection(ip, pubkey)) {
+                    SN_LOG(info, "Successfully authenticated incoming connection from " << as_hex(pubkey) << " at " << ip);
+                    status_code = "200";
+                    status_text = "";
+                    response_vals[4 /*user-id*/] = as_hex(pubkey); // ZMQ `gets` requires a null-terminated value
+                } else {
+                    SN_LOG(info, "Authentication failed for incoming connection from " << as_hex(pubkey) << " at " << ip);
+                    status_code = "400";
+                    status_text = "Access denied";
+                }
+            }
+        }
+        else if (view(frames[2]) == AUTH_DOMAIN_CLIENT) {
+            std::string ip{view(frames[3])};
+            auto mech = view(frames[5]);
+            if (mech != "NULL") {
+                status_code = "400";
+                status_text = "Client connections require NULL authentication, not " + (std::string) mech;
+            }
+            else if (allow_connection(ip, "")) {
+                SN_LOG(info, "Accepted incoming client connection from " << ip);
+                status_code = "200";
+                status_text = "";
+            } else {
+                SN_LOG(info, "Rejected incoming client connection rejected from " << ip);
+                status_code = "400";
+                status_text = "Access denied";
+            }
+        }
+        else {
+            status_code = "400";
+            status_text = "Unknown authentication domain: " + std::string{view(frames[2])};
+        }
+
+        SN_LOG(trace, "ZAP request result: " << status_code << " " << status_text);
+
+        std::list<zmq::message_t> response;
+        for (auto &r : response_vals) response.push_back(create_message(std::move(r)));
+        send_message_parts(zap_auth, response.begin(), response.end());
+    }
+}
+
+SNNetwork::~SNNetwork() {
+    SN_LOG(info, "SNNetwork shutting down proxy thread");
+    send_control(get_control_socket(), "QUIT");
+    proxy_thread.join();
+    SN_LOG(info, "SNNetwork proxy thread has stopped");
+}
+
+void SNNetwork::connect(const std::string &pubkey, std::chrono::milliseconds keep_alive) {
```

Comment:
Are these series of commands intended to be used with blink? I didn't carefully review because there's no usage code for it yet.

Reply:
Yup.  Blink is going to pre-connect when a block is received that makes it part of an active quorum (rather than waiting for a blink tx to arrive to do the connection + key exchange handshaking).